### PR TITLE
ToolsPanel: Refine component behaviour

### LINKED ISSUE #34257: ToolsPanel: Refining the component's behaviour
## What problem does this address?
Now that the new `ToolsPanel` has been [merged](https://github.com/WordPress/gutenberg/pull/32392) there’s been a few questions or concerns raised about its UX. 

The plan has always been to iterate on its behaviour and polish the panel ahead of 5.9. This issue aims to collect comments or feedback from across several PRs, issues or channels and help focus the discussion around what improvements we wish to make.

## Issues raised to date

1. The panel isn’t collapsible. All controls can be hidden but this means the value is also reset.
2. The majority of people using the panel for the first time expect the menu to only toggle visibility of controls, not reset them
3. Combining the concept of default controls along with the “reset all” menu item causes odd or unexpected behaviour. Toggling that control’s menu item results in it being hidden whereas the “reset all” doesn’t.
4. There is no persistence for panel display between block selections e.g. If you hide padding, deselect the block, then reselect it, the padding control will be back again.

## Links to feedback / comments

- https://github.com/WordPress/gutenberg/pull/32392#pullrequestreview-725158764
- https://github.com/WordPress/gutenberg/pull/32392#pullrequestreview-698325449
- https://github.com/WordPress/gutenberg/pull/32392#issuecomment-875359706
- https://github.com/WordPress/gutenberg/pull/32392#pullrequestreview-703404232
- https://github.com/WordPress/gutenberg/pull/34063#issuecomment-902943927
- https://wptavern.com/gutenberg-11-3-introduces-dimensions-panel-adds-button-padding-support-and-speeds-up-the-inserter
- https://github.com/WordPress/gutenberg/pull/32392#discussion_r664801507
- https://github.com/WordPress/gutenberg/pull/32392#issuecomment-889855872

## Related PRs

- https://github.com/WordPress/gutenberg/pull/34107

## Current panel demo

https://user-images.githubusercontent.com/60436221/130572042-39c9f76d-20d4-4626-b4c9-166921b1ac8b.mp4

## Questions

- Does further adoption of the ToolsPanel for block support panels depend on refinements to be discussed in this issue?
- Do any current explorations around improving the Post tab of the sidebar or global styles introduce any new requirements?

---

cc/ @jasmussen @shaunandrews @mtias @paaljoachim @ciampo 

I hope you don't mind the pings on this but thought it better to cast the net wide for early thoughts and feedback.

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Fixes: https://github.com/WordPress/gutenberg/issues/34257

Depends on:
- https://github.com/WordPress/gutenberg/pull/34632

## Description

- Updates the `ToolsPanel` to improve its UX following the [suggestions outlined](https://github.com/WordPress/gutenberg/issues/34257#issuecomment-908554545) on the corresponding issue.
- Adds min-width to `ToolsPanel` dropdown menu
- Adds aria-labels to menu items

## How has this been tested?

Manually:

* Tested using the group block with padding and margin support
* Tested using changes from [34065](https://github.com/WordPress/gutenberg/pull/34065) and its injected min-height control for the cover block
* Manually updated storybook story to test various configurations

#### Test instructions
1. For testing, update the Group block's config to opt into padding and margin support, enabling one as a default control. 
```
		"spacing": {
			"padding": true,
			"margin": true,
			"__experimentalDefaultControls": {
				"padding": true
			}
		},
```
2. Depending on your theme, you'll likely also need to enable the margin support UI by setting the `settings.spacing.customMargin` property to `true` in your theme.json file.
3. Add a group block in the post editor and select it
4. Explore the behaviour of the ToolsPanel via the block's "Dimensions" panel.
5. Fire up native and confirm that the `Stepper` component displays its check icon correctly. Any `UnitControl` on native will show this `Stepper` component, one easy block to test with on native is the Cover block.

#### Cover Block with cherry-picked commit from #34065
https://user-images.githubusercontent.com/60436221/132173157-a41ff448-2b83-43bb-ad54-877839013e25.mp4

#### Storybook
 https://user-images.githubusercontent.com/60436221/132362438-1a9ac63d-841d-4e62-ada9-d80b06b51e1e.mp4

## Types of changes
Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by aaronrobertshaw** [comment]
It's worth noting here that the current iteration actually sets `disabled={ true }` on a default control's menu item when it does not have a customized value. In the mockup, the text for the item is normal and only the icon greyed out. Clicking the item would do nothing.  I think disabled in this context with the separation of default and optional controls makes sense.

Open to other ideas or opinions though.

**Comment by jasmussen** [comment]
Thanks for tackling this one!

I had a bit of a challenge getting things to work as your recording; my Cover block still has a separate min-height panel, and I couldn't get margin to show despite trying a few different themes, and updating the Group block.json file. However I did see this:

hey

Some good steps here: the default panel is separate, I can reset it using the minus, I can reset all, and the check clears and removes non-default options.

We need the new icons from the mockup, these:

Screenshot 2021-09-06 at 10 51 41

Updated `check`:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M16.7 7.1l-6.3 8.5-3.3-2.5-.9 1.2 4.5 3.4L17.9 8z" /></svg>
```

New `reset`:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M7 11.5h10V13H7z" /></svg>
```

The new check is actually a bit smaller than the existing one:

Screenshot 2021-09-06 at 10 55 18

I see that we already have a minus, I had forgotten that. Do you know if/where it's used at the moment? We might want to rename it to "reset" and replace it, depending on where and how much it's used.

I'm not sure whether we want a new `check-small` icon, or whether we just replace the existing one. I'm tempted to do the former. 

Let me know if you'd like me to add these separately.

File: packages/components/src/tools-panel/tools-panel-header/component.js [hunk 1]
```diff
@@ -27,35 +41,87 @@ const ToolsPanelHeader = ( props, forwardedRef ) => {
...
+
 	return (
 		<h2 { ...headerProps } ref={ forwardedRef }>
 			{ labelText }
 			{ hasMenuItems && (
-				<DropdownMenu icon={ moreVertical } label={ labelText }>
+				<DropdownMenu
+					icon={ moreVertical }
+					label={ labelText }
+					menuProps={ { style: { minWidth: '200px' } } }
```
Comment (ciampo) [inline packages/components/src/tools-panel/tools-panel-header/component.js]: Ideally we'd like to avoid applying CSS via the inline `style` prop — the preferred option would be to pass them via a `className`

File: packages/components/src/tools-panel/tools-panel-header/component.js [hunk 2]
```diff
@@ -27,35 +41,87 @@ const ToolsPanelHeader = ( props, forwardedRef ) => {
...
-												} }
-												role="menuitemcheckbox"
-											>
-												{ label }
-											</MenuItem>
-										);
-									}
-								) }
-							</MenuGroup>
+							{ !! defaultItems?.length && (
```
Comment (ciampo) [inline packages/components/src/tools-panel/tools-panel-header/component.js]: This block is starting to get quite hard to parse, have you considered extracting it to a separate component? (it could live in this same file and it wouldn't need to be exported).

Also, have you considered [using ternaries instead of the `&&` operator](https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx)?

File: packages/components/src/tools-panel/tools-panel-header/component.js [hunk 3]
```diff
@@ -27,35 +41,87 @@ const ToolsPanelHeader = ( props, forwardedRef ) => {
...
+												? minus
+												: check;
+
+											const itemLabel = hasValue
+												? sprintf(
+														// translators: %s: The name of the control being reset e.g. "Padding".
+														__( 'Reset %s' ),
+														label
+												  )
+												: undefined;
```
Comment (ciampo) [inline packages/components/src/tools-panel/tools-panel-header/component.js]: Nit: should this piece of logic also be moved to the `getAriaLabel` function? It seems weird to use that function only for optional items

File: packages/components/src/tools-panel/tools-panel/hook.js [hunk 4]
```diff
@@ -93,10 +119,11 @@ export function useToolsPanel( props ) {
 		}
 
 		// Turn off display of all non-default items.
-		const resetMenuItems = {};
+		const resetMenuItems = { default: {}, optional: {} };
 
 		panelItems.forEach( ( { label, isShownByDefault } ) => {
-			resetMenuItems[ label ] = !! isShownByDefault;
+			const group = isShownByDefault ? 'default' : 'optional';
+			resetMenuItems[ group ][ label ] = false;
```
Comment (ciampo) [inline packages/components/src/tools-panel/tools-panel/hook.js]: Small nit: since this same logic is repeated earlier in this same file, I wonder if it could be extracted to improve the code's resistance to future changes (even by just creating a function private to this file)

**Review by ciampo — COMMENTED** [review]
Thank you for starting to work on this, @aaronrobertshaw !

I've left a few initial inline comments, and I'm going to follow the progress of this PR in the upcoming days

File: packages/components/src/tools-panel/tools-panel-header/component.js (hunk 1 above)
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/tools-panel-header/component.js]: I was under the impression I needed to avoid adding a stylesheet for components and didn't see how I could use styled-components within the `DropdownMenu` here. I'll revisit this as I work through the other suggestions. Thanks.

File: packages/components/src/tools-panel/tools-panel-header/component.js (hunk 2 above)
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/tools-panel-header/component.js]: > This block is starting to get quite hard to parse, have you considered extracting it to a separate component? 

Agreed. I've wavered on extracting the `MenuItem`s to their own components previously but it didn't reduce the code much. I'll look at refactoring the menu groups out into separate components instead.

> Also, have you considered using ternaries instead of the && operator?

In this instance, the use of `!!` was to specifically guard against the issue highlighted by that link. The `&&` operator still appears to be the prevailing style within the codebase. I understand that using a ternary makes missing the `!!` a non-issue. Is this something that should be added to the Gutenberg coding guidelines?

File: packages/components/src/tools-panel/tools-panel-header/component.js (hunk 3 above)
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/tools-panel-header/component.js]: It does but so did adding extra conditionals within `getAriaLabel` along with passing extra args. In the end, I went with the simplest version of both the function and its use. I don't feel strongly about it, so happy to change as requested.

File: packages/components/src/tools-panel/tools-panel-header/component.js (hunk 2 above)
Comment (ciampo) [reply packages/components/src/tools-panel/tools-panel-header/component.js]: >  I'll look at refactoring the menu groups out into separate components instead.

Sounds good! Thank you :)

> In this instance, the use of !! was to specifically guard against the issue highlighted by that link. The && operator still appears to be the prevailing style within the codebase. I understand that using a ternary makes missing the !! a non-issue. Is this something that should be added to the Gutenberg coding guidelines?

Thank you for the explanation about using `!!`. I'm not sure if this should be added to the coding guidelines or even as an ESLint rule — that is something that @gziolo may advise about!

File: packages/components/src/tools-panel/tools-panel-header/component.js (hunk 1 above)
Comment (ciampo) [reply packages/components/src/tools-panel/tools-panel-header/component.js]: You could define styles in the `styles.js` file, and create a className by using the `cx` function in the hook.js file

<details>

<summary>Something like this (expand):</summary>

```diff
diff --git a/packages/components/src/tools-panel/styles.js b/packages/components/src/tools-panel/styles.js
index d9c7a985c0..c4b644fe9c 100644
--- a/packages/components/src/tools-panel/styles.js
+++ b/packages/components/src/tools-panel/styles.js
@@ -46,6 +46,10 @@ export const ToolsPanelHeader = css`
 	}
 `;
 
+export const DropdownMenu = css`
+	min-width: 200px;
+`;
+
 export const ToolsPanelItem = css`
 	grid-column: span 2;
 
diff --git a/packages/components/src/tools-panel/tools-panel-header/component.js b/packages/components/src/tools-panel/tools-panel-header/component.js
index 73af46f2be..5ff8304318 100644
--- a/packages/components/src/tools-panel/tools-panel-header/component.js
+++ b/packages/components/src/tools-panel/tools-panel-header/component.js
@@ -34,6 +34,7 @@ const ToolsPanelHeader = ( props, forwardedRef ) => {
 		menuItems,
 		resetAll,
 		toggleItem,
+		dropdownMenuClassName,
 		...headerProps
 	} = useToolsPanelHeader( props );
 
@@ -51,7 +52,7 @@ const ToolsPanelHeader = ( props, forwardedRef ) => {
 				<DropdownMenu
 					icon={ moreVertical }
 					label={ labelText }
-					menuProps={ { style: { minWidth: '200px' } } }
+					menuProps={ { className: dropdownMenuClassName } }
 				>
 					{ ( { onClose } ) => (
 						<>
diff --git a/packages/components/src/tools-panel/tools-panel-header/hook.js b/packages/components/src/tools-panel/tools-panel-header/hook.js
index c903c24725..73f7fe19fe 100644
--- a/packages/components/src/tools-panel/tools-panel-header/hook.js
+++ b/packages/components/src/tools-panel/tools-panel-header/hook.js
@@ -22,6 +22,10 @@ export function useToolsPanelHeader( props ) {
 		return cx( styles.ToolsPanelHeader, className );
 	}, [ className ] );
 
+	const dropdownMenuClassName = useMemo( () => {
+		return cx( styles.DropdownMenu );
+	}, [] );
+
 	const { menuItems, hasMenuItems } = useToolsPanelContext();
 
 	return {
@@ -29,5 +33,6 @@ export function useToolsPanelHeader( props ) {
 		hasMenuItems,
 		menuItems,
 		className: classes,
+		dropdownMenuClassName,
 	};
 }
```
</details>

**Comment by ciampo** [comment]
>I had a bit of a challenge getting things to work as your recording; my Cover block still has a separate min-height panel, and I couldn't get margin to show despite trying a few different themes, and updating the Group block.json file

Just wanted to flag that I also had the same issues when trying to follow the instructions in the PR.

I also tried to use Storybook, but visiting the ToolsPanel story at the moment generates an error:

Screenshot 2021-09-07 at 10 35 19

File: packages/components/src/tools-panel/tools-panel-header/component.js (hunk 1 above)
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/tools-panel-header/component.js]: Brilliant, thanks! That certainly makes it obvious how it all fits together. 🙇 

I'll add those suggestions tomorrow.

Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/tools-panel-header/component.js]: ✅ Done.

**Comment by aaronrobertshaw** [comment]
Thanks @jasmussen for your time and effort as always 👍 

> I had a bit of a challenge getting things to work as your recording; my Cover block still has a separate min-height panel, and I couldn't get margin to show despite trying a few different themes, and updating the Group block.json file. However I did see this:

The theme I was testing with already had the `customMargin` setting turned on enabling the UI. I believe this was the key step missing, so I've updated the PR test instructions accordingly.

To test with the cover block's min-height control injected via the proposed inspector controls slots, you'd need to apply commits from https://github.com/WordPress/gutenberg/pull/34065 on top of this PR. I did this to test the updated ToolsPanel with a SlotFill injected control, I didn't intend for you to need to do it at this point in time.

I fixed the issue with the ToolsPanel story that @ciampo highlighted so that is the easiest option for getting your hands on the updated panel now.

> We need the new icons from the mockup

Awesome, thank you. I've added the new icons.

> I see that we already have a minus, I had forgotten that. Do you know if/where it's used at the moment? We might want to rename it to "reset" and replace it, depending on where and how much it's used.

At a quick glance, it only appears that the minus icon is used for the iOS stepper.

> I'm not sure whether we want a new check-small icon, or whether we just replace the existing one. I'm tempted to do the former.

I added the icon as a new check icon rather than replace the existing one.

**Comment by aaronrobertshaw** [comment]
I'm happy to update the check and reset icons however needed. Including handling them via a separate PR if the extra changes to existing icons are preferred.

**Comment by jasmussen** [comment]
Alright, I promised I'd return with more info on the checks/reset icons:

- The check is not used widely enough that it's worth keeping a larger version. So let's replace the existing `check` with the new smaller vectors, instead of adding a new one.
- The existing `minus` icon was a Dashicon, and not part of the core Figma file. So let's replace the old vectors with the new (remember to update the viewBox from "-2 -2 24 24" to "0 0 24 24"), and rename it from `minus` to `reset`. This likely requires a breaking change note for the icon component.

I've updated the core Figma file with the new icons:

Screenshot 2021-09-08 at 12 25 25

Let me know if that works!

**Review by ciampo — COMMENTED** [review]
From a quick look, code changes LGTM! I also tried the component in Storybook and it seemed to work as intended.

I'll let @jasmussen give the final seal of approval ✅

**Comment by apeatling** [comment]
This is definitely an improvement, and should be merged. I do want to note that I think there's still too much assumed knowledge wrapped up in how this works. For example this first experience for someone who has never seen this before:



I don't think it's clear what the difference between the minus and checkbox really means. It's assuming I know that the padding is a default control and therefore can't be "unchecked". It's a tough one when only icons are being used with no labels indicating which controls are the default and can't be removed. I don't have any good suggestions, but wanted to note this.

**Review by apeatling — COMMENTED** [review]
This is definitely an improvement, and should be merged. 

I do want to note that I think there's still too much assumed knowledge wrapped up in how this works. For example this first experience for someone who has never seen this before:



I don't think it's clear what the difference in expected action between the minus and check icon is. It's assuming I know that the padding is a default control and therefore can't be "unchecked". 

It's a tough one when only icons are being used with no labels indicating which controls are the default and therefore can't be removed. I don't have any good suggestions, but wanted to note this as something to think about further.

**Comment by aaronrobertshaw** [comment]
✅ The `check` and `minus` (now `reset`) icons have now been replaced with the supplied vectors
✅ Note about breaking change for `minus` icon added to changelog
✅ Confirmed native `Stepper` component still works after switching to `reset` icon
✅ Rebased this on top of trunk

**Comment by aaronrobertshaw** [comment]
> I do want to note that I think there's still too much assumed knowledge wrapped up in how this works.

I agree with this. Solving it is a tricky one though.

> It's a tough one when only icons are being used with no labels indicating which controls are the default and therefore can't be removed. I don't have any good suggestions, but wanted to note this as something to think about further.

Adding some tooltips to the panel's menu items was trialled but it made the UI incredibly busy, undermining any improvements to the UX.

> This is definitely an improvement, and should be merged.

I'd be happy for this to land once @jasmussen has a chance to approve the PR in its current state. 

We can then iterate on this further in follow up PRs. Merging this will also allow for the `ToolsPanel` conversion to TypeScript to continue with less conflicts.

**Comment by jasmussen** [comment]
Took it for a quick spin, and the paddings, margins, and behavior looks right to me:

tools

> I do want to note that I think there's still too much assumed knowledge wrapped up in how this works. 

The goal of the interface is to provide additional tools first and foremost, and secondarily to provide an additional and intentionally de-emphasized affordance for clearing values. Knowing that this interface can and will be iterated upon in the future, it feels important to start as simple as possible.

In that vein, let's definitely ship this one. And if we find the need to adjust further, we will. Thanks Aaron!

**Review by jasmussen — APPROVED** [review]
